--- old/src/java.base/share/native/libjli/args.c	2019-12-06 15:12:10.223425900 -0500
+++ new/src/java.base/share/native/libjli/args.c	2019-12-06 15:12:09.494362900 -0500
@@ -130,6 +130,8 @@
             }
         } else if (JLI_StrCmp(arg, "--disable-@files") == 0) {
             stopExpansion = JNI_TRUE;
+        } else if (JLI_StrCCmp(arg, "--module=") == 0) {
+            idx = argsCount;
         }
     } else {
         if (!expectingNoDashArg) {
